File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root:
            return False
        return root.val == subRoot.val and self.isSame(root,subRoot) or self.isSubtree(root.left,subRoot) or self.isSubtree(root.right,subRoot)
            
    
    def isSame(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and q or not q and p:
            return False
        elif not p and not q:
            return True
        return p.val==q.val and self.isSame(p.left,q.left) and self.isSame(p.right,q.right)
        

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
unordered_map<int,bool> mp;
TreeNode* ans;
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        srch(root,p);
        srch(root,q);
        return ans;
    }
    void srch(TreeNode* root, TreeNode* p){
        if(mp[root->val])
            ans=root;
        else
            mp[root->val]=true;
        if(root->val==p->val)
            return;
        else if(root->val<p->val)
            srch(root->right,p);
        else
            srch(root->left,p);
    }
};